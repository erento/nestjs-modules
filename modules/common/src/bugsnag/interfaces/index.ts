export enum BugsnagSeverity {
    INFO = 'info',
    WARNING = 'warn',
    ERROR = 'error',
}
