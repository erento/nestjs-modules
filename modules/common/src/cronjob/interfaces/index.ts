export interface CronJobMethodReference {
    target: Object;
    methodName: string;
}
