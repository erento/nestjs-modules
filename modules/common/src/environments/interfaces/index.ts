export enum EnvironmentType {
    PROD = 'production',
    BETA = 'beta', // will be treated as production
    DEV = 'development',
    TEST = 'test',
}
