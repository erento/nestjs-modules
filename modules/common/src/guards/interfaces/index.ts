export enum AuthorizationType {
    service = 'service',
    admin = 'admin',
    seller = 'seller',
}
