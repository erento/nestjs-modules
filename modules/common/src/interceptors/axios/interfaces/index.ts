export type AxiosFulfilledInterceptor<T> = (value: T) => T | Promise<T>;
export type AxiosRejectedInterceptor = (error: any) => any;
