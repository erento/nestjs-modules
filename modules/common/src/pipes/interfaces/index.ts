export interface PaginationParams {
    page: number;
    size: number;
}

export interface PaginatedResponse<T> {
    page: number;
    pages: number;
    size: number;
    total: number;
    results: T[];
}

export interface OffsetAndLimit {
    offset?: number;
    limit?: number;
}
