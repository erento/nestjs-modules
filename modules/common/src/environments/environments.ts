import {EnvironmentType} from './interfaces';

export class Environments {
    public static getEnv (): EnvironmentType {
        return process.env.NODE_ENV === EnvironmentType.PROD ?
            EnvironmentType.PROD :
            process.env.NODE_ENV === EnvironmentType.TEST ?
                EnvironmentType.TEST :
                process.env.NODE_ENV === EnvironmentType.BETA ?
                    EnvironmentType.BETA :
                    EnvironmentType.DEV;
    }

    public static isProd (): boolean {
        return this.getEnv() === EnvironmentType.PROD || this.getEnv() === EnvironmentType.BETA;
    }

    public static isDev (): boolean {
        return this.getEnv() === EnvironmentType.DEV;
    }

    public static isTest (): boolean {
        return this.getEnv() === EnvironmentType.TEST;
    }

    public static getPackageJson (): any {
        // tslint:disable-next-line no-var-requires no-require-imports
        return require(process.cwd() + '/package.json');
    }

    public static getReleaseStage (): EnvironmentType {
        return this.getEnv();
    }

    public static getVersion (): string {
        return process.env.APP_VERSION || EnvironmentType.DEV;
    }
}
