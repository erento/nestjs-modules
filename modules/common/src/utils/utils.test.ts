import {EnvironmentType} from '../environments/interfaces';
import {
    arrayDifference,
    arrayIntersection,
    batchArray,
    cleanNullValues,
    createBuffer,
    deepCopyObj,
    fillArrayWithIndices,
    getDateStringOrUndefined,
    getPropertyOrFallback,
    getPropertyOrFallbackOrUndefined,
    getPropertyOrUndefined,
    getTrueOrUndefined,
    isUuid,
    oneLine,
    parseDateOrUndefined,
    replaceEmptyStringValuesWithNull,
    requiredEnvVariable
} from './utils';

describe('utils', (): void => {
    describe('buffer', (): void => {
        test('create', (): void => {
            const base64: string = 'YWhveQ==';
            const buffer: Buffer | undefined = createBuffer(base64);

            expect(buffer).toBeInstanceOf(Buffer);
            expect(buffer.toString()).toMatchSnapshot();
        });

        test('create with wrong data', (): void => {
            expect(createBuffer('not a base 64')).toBeInstanceOf(Buffer);
            expect((): Buffer => createBuffer(<any> null)).toThrowErrorMatchingSnapshot();
            expect((): Buffer => createBuffer(<any> undefined)).toThrowErrorMatchingSnapshot();
        });
    });

    describe('array', (): void => {
        test('difference', (): void => {
            expect(arrayDifference(['a', 'b'], ['b', 'c'])).toMatchSnapshot();
            expect(arrayDifference(['a', 'b'], [])).toMatchSnapshot();
            expect(arrayDifference(['a', 'b'], ['a', 'b'])).toMatchSnapshot();
        });

        test('intersection', (): void => {
            expect(arrayIntersection(['a', 'b'], ['b', 'c'])).toMatchSnapshot();
            expect(arrayIntersection(['a', 'b'], ['c', 'd'])).toMatchSnapshot();
            expect(arrayIntersection([{a: 'a'}], [{a: 'a'}])).toMatchSnapshot();

            const value: any = {b: 'b'};
            expect(arrayIntersection(['a', value, 'b'], [value, 'b'])).toMatchSnapshot();
        });

        test('intersection is unique', (): void => {
            expect(arrayIntersection(['a', 12], ['c', 12, 'd', 12])).toMatchSnapshot();
        });
    });

    describe('object', (): void => {
        test('copy', (): void => {
            const objA: any = {a: {a: 1}, b: {b: 1}, c: 1};
            const copy: any = deepCopyObj(objA);
            expect(copy).toMatchSnapshot();
            objA.a.a = 'changed value';
            expect(copy).toMatchSnapshot();
        });

        test('copy not an object', (): void => {
            expect(deepCopyObj(undefined)).toMatchSnapshot();
            expect(deepCopyObj(null)).toMatchSnapshot();
            expect(deepCopyObj(NaN)).toMatchSnapshot();
            expect(deepCopyObj('')).toMatchSnapshot();
            expect(deepCopyObj('some string')).toMatchSnapshot();
            expect(deepCopyObj(1)).toMatchSnapshot();
        });
    });

    describe('undefined or value', (): void => {
        test('get any value or undefined', (): void => {
            expect(getPropertyOrUndefined('some test')).toMatchSnapshot();
            expect(getPropertyOrUndefined({some: 'thing', other: 'thing'})).toMatchSnapshot();
            expect(getPropertyOrUndefined(null)).toMatchSnapshot();
            expect(getPropertyOrUndefined(undefined)).toMatchSnapshot();
        });

        test('get date string or undefined', (): void => {
            const testEpoch: number = 1539952471893;
            expect(getDateStringOrUndefined('some test')).toMatchSnapshot();
            expect(getDateStringOrUndefined(null)).toEqual(undefined);
            expect(getDateStringOrUndefined(new Date(testEpoch))).toMatchSnapshot();
            // special case for dummy dates set as default if there is no timestamp existing
            expect(getDateStringOrUndefined(new Date(0))).toMatchSnapshot();
        });

        test('get true or undefined', (): void => {
            expect(getTrueOrUndefined(false)).toMatchSnapshot();
            expect(getTrueOrUndefined(true)).toMatchSnapshot();
        });

        test('parse date or undefined', (): void => {
            expect(parseDateOrUndefined('xx-x-x-x')).toMatchSnapshot();
            expect(parseDateOrUndefined(undefined)).toMatchSnapshot();
            expect(parseDateOrUndefined('1977-10-11')).toMatchSnapshot();
        });
    });

    test('isUuid', (): void => {
        expect(isUuid('11111111-2222-3333-4444-555555555555')).toBe(true);
        expect(isUuid('')).toBe(false);
        expect(isUuid(null)).toBe(false);
        expect(isUuid(undefined)).toBe(false);
        expect(isUuid('123456789')).toBe(false);
    });

    test('cleanNullValues', (): void => {
        expect(cleanNullValues({test: null})).toMatchSnapshot();
        expect(cleanNullValues({test: undefined})).toMatchSnapshot();
        expect(cleanNullValues({test: 123, property: null})).toMatchSnapshot();
        expect(cleanNullValues(null)).toMatchSnapshot();
        expect(cleanNullValues(undefined)).toMatchSnapshot();
        expect(cleanNullValues({test: null, another: null})).toMatchSnapshot();
    });

    test('replaceEmptyStringValuesWithNull', (): void => {
        expect(replaceEmptyStringValuesWithNull(null)).toMatchSnapshot();
        expect(replaceEmptyStringValuesWithNull({test: '', property: undefined})).toMatchSnapshot();
        expect(replaceEmptyStringValuesWithNull({test: '', property: null})).toMatchSnapshot();
        expect(replaceEmptyStringValuesWithNull({test: '', property: null, element: 'is-good'})).toMatchSnapshot();
        expect(replaceEmptyStringValuesWithNull({test: 1, property: null, element: 'is-good'})).toMatchSnapshot();
    });

    describe('property or fallback or undefined', (): void => {
        test('return truthy properties except of empty string', (): void => {
            expect(getPropertyOrFallbackOrUndefined('some')).toEqual('some');
            expect(getPropertyOrFallbackOrUndefined(true)).toBe(true);
        });

        test('return undefined on falsy properties except of false', (): void => {
            expect(getPropertyOrFallbackOrUndefined(undefined)).toBeUndefined();
            expect(getPropertyOrFallbackOrUndefined(null)).toBeUndefined();
        });

        test('return undefined on empty string', (): void => {
            expect(getPropertyOrFallbackOrUndefined('')).toBeUndefined();
        });

        test('return false on false', (): void => {
            expect(getPropertyOrFallbackOrUndefined(false)).toBe(false);
        });

        test('return fallback if it\'s given and the property is undefined', (): void => {
            expect(getPropertyOrFallbackOrUndefined(undefined, 'some')).toBe('some');
        });

        test('return undefined if fallback is falsy', (): void => {
            expect(getPropertyOrFallbackOrUndefined(undefined, undefined)).toBeUndefined();
            expect(getPropertyOrFallbackOrUndefined(undefined, null)).toBeUndefined();
        });

        test('return undefined if fallback is empty string', (): void => {
            expect(getPropertyOrFallbackOrUndefined(undefined, '')).toBeUndefined();
        });
    });

    describe('property or fallback', (): void => {
        test('return value if it is truthy except of empty string', (): void => {
            expect(getPropertyOrFallback('some', 'thing')).toBe('some');
            expect(getPropertyOrFallback(true, false)).toBe(true);
        });

        test('return fallback on falsy properties except of false', (): void => {
            expect(getPropertyOrFallback(undefined, 'thing')).toBe('thing');
            expect(getPropertyOrFallback(null, 'thing')).toBe('thing');
        });

        test('return fallback on empty string', (): void => {
            expect(getPropertyOrFallback('', 'thing')).toBe('thing');
        });

        test('return value on false', (): void => {
            expect(getPropertyOrFallback(false, true)).toBe(false);
        });
    });

    describe('fillArrayWithIndices', (): void => {
        test('should return an array with indices providing a length', (): void => {
            expect(fillArrayWithIndices(5)).toMatchSnapshot();
        });

        test('should return an empty array on zero length', (): void => {
            expect(fillArrayWithIndices(0)).toMatchSnapshot();
        });

        test('should return an empty array on undefined length', (): void => {
            expect(fillArrayWithIndices()).toMatchSnapshot();
        });
    });

    describe('batchArray', (): void => {
        test('empty array', (): void => {
            expect(batchArray<string>([], 2)).toEqual([]);
        });

        test('longer array', (): void => {
            expect(batchArray<string>(['a', 'b', 'c', 'd', 'e'], 2)).toEqual([
                ['a', 'b'],
                ['c', 'd'],
                ['e'],
            ]);
        });
    });

    describe('requiredEnvVariable', (): void => {
        test('it returns process.env variable, if it exists', (): void => {
            process.env.__TESTING_PROPERTY = 'some-value';

            expect(requiredEnvVariable('__TESTING_PROPERTY')).toBe('some-value');
        });

        test('it throws for missing environment variables', (): void => {
            process.env.NODE_ENV = EnvironmentType.BETA;
            expect((): string => requiredEnvVariable('__NOT_EXISTING')).toThrowError('Missing environment variable "__NOT_EXISTING"!');
            process.env.NODE_ENV = EnvironmentType.TEST;
        });
    });

    describe('oneLine', (): void => {
        test('it converts multiline strings to one line and trims them', (): void => {
            expect(oneLine(``)).toEqual('');
            expect(oneLine(`      Hello,\n            World           `)).toEqual('Hello, World');
            expect(oneLine(`\nSELECT *\nFROM "user"\n WHERE field = 'value';`)).toEqual('SELECT * FROM "user" WHERE field = \'value\';');
            expect(oneLine(`
                SELECT *
                FROM "user"
                WHERE field = 'value'
                    AND other_field = 'other_value        with_spaces'
                LIMIT 10;
            `)).toEqual('SELECT * FROM "user" WHERE field = \'value\' AND other_field = \'other_value        with_spaces\' LIMIT 10;');
            expect(oneLine(`UPDATE "user" SET field='multi word value with any amount  of    spaces'`)).toEqual(`UPDATE "user" SET field='multi word value with any amount  of    spaces'`);
            expect(oneLine(`UPDATE "user" SET\n field='multi word value with any amount  of    spaces'`)).toEqual(`UPDATE "user" SET field='multi word value with any amount  of    spaces'`);
        });
    });
});
