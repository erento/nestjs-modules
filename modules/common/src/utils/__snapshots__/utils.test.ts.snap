// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils array difference 1`] = `
Array [
  "a",
  "c",
]
`;

exports[`utils array difference 2`] = `
Array [
  "a",
  "b",
]
`;

exports[`utils array difference 3`] = `Array []`;

exports[`utils array intersection 1`] = `
Array [
  "b",
]
`;

exports[`utils array intersection 2`] = `Array []`;

exports[`utils array intersection 3`] = `Array []`;

exports[`utils array intersection 4`] = `
Array [
  Object {
    "b": "b",
  },
  "b",
]
`;

exports[`utils array intersection is unique 1`] = `
Array [
  12,
]
`;

exports[`utils buffer create 1`] = `"ahoy"`;

exports[`utils buffer create with wrong data 1`] = `"Cannot create a buffer."`;

exports[`utils buffer create with wrong data 2`] = `"Cannot create a buffer."`;

exports[`utils cleanNullValues 1`] = `undefined`;

exports[`utils cleanNullValues 2`] = `
Object {
  "test": undefined,
}
`;

exports[`utils cleanNullValues 3`] = `
Object {
  "test": 123,
}
`;

exports[`utils cleanNullValues 4`] = `undefined`;

exports[`utils cleanNullValues 5`] = `undefined`;

exports[`utils cleanNullValues 6`] = `undefined`;

exports[`utils fillArrayWithIndices should return an array with indices providing a length 1`] = `
Array [
  0,
  1,
  2,
  3,
  4,
]
`;

exports[`utils fillArrayWithIndices should return an empty array on undefined length 1`] = `Array []`;

exports[`utils fillArrayWithIndices should return an empty array on zero length 1`] = `Array []`;

exports[`utils object copy 1`] = `
Object {
  "a": Object {
    "a": 1,
  },
  "b": Object {
    "b": 1,
  },
  "c": 1,
}
`;

exports[`utils object copy 2`] = `
Object {
  "a": Object {
    "a": 1,
  },
  "b": Object {
    "b": 1,
  },
  "c": 1,
}
`;

exports[`utils object copy not an object 1`] = `undefined`;

exports[`utils object copy not an object 2`] = `null`;

exports[`utils object copy not an object 3`] = `null`;

exports[`utils object copy not an object 4`] = `""`;

exports[`utils object copy not an object 5`] = `"some string"`;

exports[`utils object copy not an object 6`] = `1`;

exports[`utils replaceEmptyStringValuesWithNull 1`] = `null`;

exports[`utils replaceEmptyStringValuesWithNull 2`] = `
Object {
  "property": null,
  "test": null,
}
`;

exports[`utils replaceEmptyStringValuesWithNull 3`] = `
Object {
  "property": null,
  "test": null,
}
`;

exports[`utils replaceEmptyStringValuesWithNull 4`] = `
Object {
  "element": "is-good",
  "property": null,
  "test": null,
}
`;

exports[`utils replaceEmptyStringValuesWithNull 5`] = `
Object {
  "element": "is-good",
  "property": null,
  "test": 1,
}
`;

exports[`utils undefined or value get any value or undefined 1`] = `"some test"`;

exports[`utils undefined or value get any value or undefined 2`] = `
Object {
  "other": "thing",
  "some": "thing",
}
`;

exports[`utils undefined or value get any value or undefined 3`] = `undefined`;

exports[`utils undefined or value get any value or undefined 4`] = `undefined`;

exports[`utils undefined or value get date string or undefined 1`] = `undefined`;

exports[`utils undefined or value get date string or undefined 2`] = `"2018-10-19T12:34:31.893Z"`;

exports[`utils undefined or value get date string or undefined 3`] = `undefined`;

exports[`utils undefined or value get true or undefined 1`] = `undefined`;

exports[`utils undefined or value get true or undefined 2`] = `true`;

exports[`utils undefined or value parse date or undefined 1`] = `undefined`;

exports[`utils undefined or value parse date or undefined 2`] = `undefined`;

exports[`utils undefined or value parse date or undefined 3`] = `1977-10-11T00:00:00.000Z`;
